openssl-dynamic/src/main/c/ocsp.c

```diff
@@ -0,0 +1,216 @@
+    // error and success to OpenSSL. We're going to do something naughty to
+    // align OpenSSL's and BoringSSL's APIs and simply tell OpenSSL to use
+    // a stubbed callback function that is always saying that things are OK.
+    //
+    error = SSL_CTX_set_tlsext_status_arg(c->ctx, (client ? &OCSP_CLIENT_ACK : &OCSP_SERVER_ACK));
+    if (error <= 0L) {
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_arg() failed");
+        return JNI_FALSE;
+    }
+
+    error = SSL_CTX_set_tlsext_status_cb(c->ctx, openssl_ocsp_callback);
+    if (error <= 0L) {
+        tcn_ThrowException(e, "SSL_CTX_set_tlsext_status_cb() failed");
+        return JNI_FALSE;
+    }
+
+    return JNI_TRUE;
+#endif
+}
+
+/**
+ * Disables OCSP stapling for the given SSLContext
+ */
+TCN_IMPLEMENT_CALL(void, SSLContext, disableOcsp)(TCN_STDARGS, jlong ctx) {
+// This does nothing in BoringSSL
+#if !defined(TCN_NO_OCSP) && !defined(OPENSSL_IS_BORINGSSL)
+    tcn_ssl_ctxt_t *c = J2P(ctx, tcn_ssl_ctxt_t *);
+    if (c != NULL) {
+      SSL_CTX_set_tlsext_status_cb(c->ctx, NULL);
+      SSL_CTX_set_tlsext_status_arg(c->ctx, NULL);
+    }
+#endif
+}
+
+/**
+ * Enables OCSP stapling on the SSLEngine.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, enableOcsp)(TCN_STDARGS, jlong ssl) {
+#ifdef TCN_NO_OCSP
+  return JNI_FALSE;
+
+#else
+    SSL *ssl_ = J2P(ssl, SSL *);
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return JNI_FALSE;
+    }
+#ifdef OPENSSL_IS_BORINGSSL
+    SSL_enable_ocsp_stapling(ssl_);
+    return JNI_TRUE;
+
+#else
+    return (SSL_set_tlsext_status_type(ssl_, TLSEXT_STATUSTYPE_ocsp) == 1L) ? JNI_TRUE : JNI_FALSE;
+
+#endif /* OPENSSL_IS_BORINGSSL */
+#endif /* TCN_NO_OCSP */
+}
+
+/**
+ * Server: Sets OCSP response bytes.
+ */
+TCN_IMPLEMENT_CALL(jboolean, SSL, setOcspResponse)(TCN_STDARGS, jlong ssl, jbyteArray response) {
+#ifdef TCN_NO_OCSP
+    return JNI_FALSE;
+
+#else
+    SSL *ssl_ = J2P(ssl, SSL *);
+    if (ssl_ == NULL) {
+        tcn_ThrowException(e, "ssl is null");
+        return JNI_FALSE;
+    }
+
+    jsize length = (*e)->GetArrayLength(e, response);
+    if (length == 0) {
+        return JNI_FALSE;
+    }
+
+#ifdef OPENSSL_IS_BORINGSSL
+    const uint8_t value[length];
+    (*e)->GetByteArrayRegion(e, response, 0, length, (jbyte*)&value);
+
+    return (SSL_set_ocsp_response(ssl_, &value, length) != 0 ? JNI_TRUE : JNI_FALSE);
+
+#else
+    //
+    // ATTENTION: This took a while to figure out but OpenSSL wants to (and will)
+    // free() this pointer on its own. Give it something it can free or it will crash.
+    //
+    unsigned char *value = OPENSSL_malloc(sizeof(unsigned char) * length);
+    if (value == NULL) {
+        tcn_ThrowException(e, "OPENSSL_malloc() returned null");
+        return JNI_FALSE;
+    }
+
+    (*e)->GetByteArrayRegion(e, response, 0, length, (jbyte*)value);
+    long error = SSL_set_tlsext_status_ocsp_resp(ssl_, value, length);
```

Comment:
nit: remove `error` and just inline the call in the `if` statement `if (SSL_set_tlsext_status_ocsp_resp(ssl_, value, length) != 1L) { .. }`